spec/bucket_store_integration_spec.rb

```diff
@@ -24,55 +26,162 @@
+      context "when handling streaming uploads / downloads" do
+        let(:temp_filename) { "temp-file" }
+        let(:buffer_filename) { "in-memory-buffer" }
+        let(:input_temp_file) do
+          Tempfile.new.tap do |file|
+            file.write(Random.base64(1024))
+            file.rewind
+          end
+        end
+        let(:output_temp_file) do
+          Tempfile.new
+        end
+
+        let(:input_in_memory_buffer) do
+          StringIO.new(Random.base64(1024))
+        end
+        let(:output_in_memory_buffer) do
+          StringIO.new
+        end
+
+        before do
+          described_class.for("#{base_bucket_uri}/#{temp_filename}").
+            stream.
+            upload!(file: input_temp_file)
+          input_temp_file.rewind
+
+          described_class.for("#{base_bucket_uri}/#{buffer_filename}").
+            stream.
+            upload!(file: input_in_memory_buffer)
+          input_in_memory_buffer.rewind
+        end
+
+        after do
+          input_temp_file.close
+          output_temp_file.close
+        end
+
+        it "all file are listable on the store" do
+          expect(described_class.for(base_bucket_uri.to_s).list.to_a.size).to eq(2)
+        end
+
+        it "files are deletable on the store" do
+          described_class.for(base_bucket_uri.to_s).list.each do |key|
+            described_class.for(key).delete!
+          end
+          expect(described_class.for(base_bucket_uri.to_s).list.to_a.size).to eq(0)
+        end
+
+        it "files are downloadable from the store" do
+          described_class.for("#{base_bucket_uri}/#{temp_filename}").
+            stream.
+            download(file: output_temp_file)
+          output_temp_file.rewind
+
+          expect(output_temp_file.read).to eq(input_temp_file.read)
+
+          described_class.for("#{base_bucket_uri}/#{buffer_filename}").
+            stream.
+            download(file: output_in_memory_buffer)
+          output_in_memory_buffer.rewind
+
+          expect(output_in_memory_buffer.read).to eq(input_in_memory_buffer.read)
+        end
+
+        context "when the files are binary" do
+          let(:input_temp_file) do
+            Tempfile.new.binmode.tap do |file|
+              file.write(Random.bytes(1024))
+              file.rewind
+            end
+          end
+          let(:output_temp_file) do
+            Tempfile.new.binmode
+          end
+
+          let(:input_in_memory_buffer) do
+            StringIO.new.binmode.tap do |buffer|
+              buffer.write(Random.bytes(1024))
+              buffer.rewind
+            end
+          end
+          let(:output_in_memory_buffer) do
+            StringIO.new.binmode
+          end
+
+          it "files are downloadable from the store" do
+            described_class.for("#{base_bucket_uri}/#{temp_filename}").
+              stream.
+              download(file: output_temp_file)
+            output_temp_file.rewind
```

Comment:
This test passes both before and after adding the different `.rewind` calls in the download methods, since its doing the rewind itself. We need to delete this `.rewind` call (and all the other ones on the output buffers in these new specs) as otherwise we are not testing the correct behaviour.